### PR TITLE
Update wfs tutorial

### PR DESCRIPTION
**Pull request by hansvancalster**
This PR adds examples to filter on the attribute table of a WFS.

If anyone knows how to query just the attribute table of a WFS (without spatial data), that would be interesting to add. Or even better: how to see the unique values of a field of the attribute table you are interested in?

### CODE REVIEW

File: content/tutorials/spatial_wfs_services/index.Rmd [hunk 1]
```diff
@@ -238,8 +238,84 @@ ggplot(bel_regions) +
...
+  GET()
+```
+
+The CQL filter format is much more human readable and easier to code:
+
+```{r}
+wfs_vrbg %>% 
+  parse_url %>% 
+  list_merge(query = list(service = "wfs",
+                          version = "2.0.0",
```
Comment (florisvdh) [inline content/tutorials/spatial_wfs_services/index.Rmd:306]: same here for `service` & `version`

**Review by florisvdh — COMMENTED** [review]
Thank you for elaborating @hansvancalster :+1:  Some general remarks:

- regarding (at least) the `wfs_vrbg`, it is at version 1.1.0, not 2.0.0 (apparently it doesn't really care, though). I suggest to set it right in the `ows4R` calls.
- the `service` and `version` arguments appear unneeded inside the `query` element, so probably easier to just drop them. (I tested that with the provinces example, but please double check for the other cases)
- best explain somewhere (probably in example 1) that you don't need `ows4R` to just build the URL with the query element and make the request. For scripts in production, this insight is important (the package does not need to be loaded).

Please don't forget to knit again after making the changes.

**Comment by florisvdh** [comment]
@hansvancalster while looking into new challenges with the result, I saw that the requirement of `service` and `version` arguments depends on the WFS. For `"https://eservices.minfin.fgov.be/arcgis/services/R2C/Regions/MapServer/WFSServer"` it seems that `service = "wfs"` is required. `version` is not.

A more difficult thing is the exotic **geometry type** of the resulting `sf` object in several cases (up to now I discovered: BWK has `CURVEPOLYGON` and `COMPOUNDCURVE` layers, and the provinces example is `MULTISURFACE`), and what you can (not) do with it. `geom_sf()` seems no problem, but e.g. `st_buffer()`/`st_join()` etc will not work on those. For the provinces examples, [this](https://github.com/r-spatial/sf/issues/748#issuecomment-389811593) suggestion works, but you need to know it! For the other ones, no success yet.

**Comment by hansvancalster** [comment]
> @hansvancalster while looking into new challenges with the result, I saw that the requirement of `service` and `version` arguments depends on the WFS. For `"https://eservices.minfin.fgov.be/arcgis/services/R2C/Regions/MapServer/WFSServer"` it seems that `service = "wfs"` is required. `version` is not.
> 
Yes, I had noticed that and will leave service = "wfs" in all queries just to be on the safe side.

**Comment by hansvancalster** [comment]
> A more difficult thing is the exotic **geometry type** of the resulting `sf` object in several cases (up to now I discovered: BWK has `CURVEPOLYGON` and `COMPOUNDCURVE` layers, and the provinces example is `MULTISURFACE`), and what you can (not) do with it. `geom_sf()` seems no problem, but e.g. `st_buffer()`/`st_join()` etc will not work on those. For the provinces examples, [this](https://github.com/r-spatial/sf/issues/748#issuecomment-389811593) suggestion works, but you need to know it! For the other ones, no success yet.

See 3dd0aa4

**Comment by florisvdh** [comment]
line 422 (or so): `wfs_bodemtypes <- "https://www.dov.vlaanderen.be/geoserver/bodemkaart/bodemtypes/wfs?"`

The `?` at the end can be dropped.

**Comment by florisvdh** [comment]
Any further ideas for the remaining cases? I'm prepared to post an issue at `sf` for this: extending `st_cast()`, perhaps also requesting out-of-the-box support for extra feature types, as has been done in https://github.com/r-spatial/sf/commit/fe70feab8375147bfa4c8299496785051c7f08e7 for `MULTICURVE`, IIUC.

``` r
suppressPackageStartupMessages(library(dplyr))
library(httr)
library(purrr)
library(sf)
#> Linking to GEOS 3.9.0, GDAL 3.1.3, PROJ 7.2.0
library(ggplot2)

getfeature_sf <- function(wfs, typename, cql_filter) {
  parse_url(wfs) %>%
    list_merge(query = list(service = "wfs",
                            request = "GetFeature",
                            typeName = typename,
                            cql_filter = cql_filter)) %>%
    build_url() %>%
    read_sf()
}

#########################
# MULTISURFACE
# solved using https://github.com/r-spatial/sf/issues/748#issuecomment-389811593
#########################
object1 <-
  getfeature_sf("https://geoservices.informatievlaanderen.be/overdrachtdiensten/VRBG/wfs",
                "VRBG:Refprv",
                "NAAM='West-Vlaanderen'")
st_geometry_type(object1)
#> [1] MULTISURFACE
#> 18 Levels: GEOMETRY POINT LINESTRING POLYGON MULTIPOINT ... TRIANGLE
ggplot(object1) + geom_sf()
```



``` r
st_buffer(object1, 10)
#> Error in CPL_geos_op("buffer", x, dist, nQ, numeric(0), logical(0)): Evaluation error: ParseException: Unknown WKB type 12.

object1 <-
  st_cast(object1, "GEOMETRYCOLLECTION")
st_geometry_type(object1)
#> [1] GEOMETRYCOLLECTION
#> 18 Levels: GEOMETRY POINT LINESTRING POLYGON MULTIPOINT ... TRIANGLE
st_buffer(object1, 10)
#> Simple feature collection with 1 feature and 7 fields
#> geometry type:  POLYGON
#> dimension:      XY
#> bbox:           xmin: 21981.38 ymin: 155918.6 xmax: 90426.92 ymax: 229729.6
#> CRS:            NA
#> # A tibble: 1 x 8
#>   gml_id   UIDN  OIDN TERRID NAAM   NISCODE NUTS2                          SHAPE
#> * <chr>   <dbl> <dbl>  <dbl> <chr>  <chr>   <chr>                      <POLYGON>
#> 1 Refprv…    14     3    351 West-… 30000   BE25  ((23188.45 197154.4, 23187.76…


#########################
# COMPOUNDCURVE
#########################
object2 <-
  getfeature_sf("https://geoservices.informatievlaanderen.be/overdrachtdiensten/BWK/wfs",
                "BWK:Hab3260",
                "OBJ=2071")
st_geometry_type(object2)
#> [1] COMPOUNDCURVE
#> 18 Levels: GEOMETRY POINT LINESTRING POLYGON MULTIPOINT ... TRIANGLE
ggplot(object2) + geom_sf()
```



``` r
st_buffer(object2, 10)
#> Error in CPL_geos_op("buffer", x, dist, nQ, numeric(0), logical(0)): Evaluation error: ParseException: Unknown WKB type 9.
object2$SHAPE
#> Geometry set for 1 feature 
#> geometry type:  COMPOUNDCURVE
#> dimension:      XY
#> bbox:           xmin: 127827 ymin: 167681.2 xmax: 127857.1 ymax: 167766.4
#> CRS:            NA
#> COMPOUNDCURVE (LINESTRING (127857.1 167681.2, 1...
st_cast(object2, "GEOMETRYCOLLECTION")
#> Error in which_sfc_col(from_cls): st_cast for COMPOUNDCURVE not supported
st_cast(object2, "GEOMETRY")
#> Error in which_sfc_col(from_cls): st_cast for COMPOUNDCURVE not supported
st_cast(object2, "LINESTRING")
#> Error in which_sfc_col(from_cls): st_cast for COMPOUNDCURVE not supported
st_collection_extract(object2, "LINESTRING")
#> Error in st_collection_extract.sf(object2, "LINESTRING"): x is of singular geometry type that is different to supplied type: LINESTRING

#########################
# CURVEPOLYGON
#########################
object3 <-
  getfeature_sf("https://geoservices.informatievlaanderen.be/overdrachtdiensten/BWK/wfs",
                "BWK:Bwkhab",
                "TAG='000098_v2018'")
st_geometry_type(object3)
#> [1] CURVEPOLYGON
#> 18 Levels: GEOMETRY POINT LINESTRING POLYGON MULTIPOINT ... TRIANGLE
ggplot(object3) + geom_sf()
```



``` r
st_buffer(object3, 10)
#> Error in CPL_geos_op("buffer", x, dist, nQ, numeric(0), logical(0)): Evaluation error: ParseException: Unknown WKB type 10.
object3$SHAPE
#> Geometry set for 1 feature 
#> geometry type:  CURVEPOLYGON
#> dimension:      XY
#> bbox:           xmin: 114681.6 ymin: 211527.9 xmax: 114738 ymax: 211584.1
#> CRS:            NA
#> CURVEPOLYGON (LINESTRING (114738 211559.6, 1147...
st_cast(object3, "GEOMETRYCOLLECTION")
#> Error in which_sfc_col(from_cls): st_cast for CURVEPOLYGON not supported
st_cast(object3, "GEOMETRY")
#> Error in which_sfc_col(from_cls): st_cast for CURVEPOLYGON not supported
st_cast(object3, "POLYGON")
#> Error in which_sfc_col(from_cls): st_cast for CURVEPOLYGON not supported
st_cast(object3, "LINESTRING")
#> Error in which_sfc_col(from_cls): st_cast for CURVEPOLYGON not supported
st_collection_extract(object3, "LINESTRING")
#> Error in st_collection_extract.sf(object3, "LINESTRING"): x is of singular geometry type that is different to supplied type: LINESTRING
```

<sup>Created on 2021-01-14 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
<details>
<summary>
Session info
</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.3 (2020-10-10)
#>  os       Linux Mint 20               
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language nl_BE:nl                    
#>  collate  nl_BE.UTF-8                 
#>  ctype    nl_BE.UTF-8                 
#>  tz       Europe/Brussels             
#>  date     2021-01-14                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.2)
#>  callr         3.5.1   2020-10-13 [1] CRAN (R 4.0.3)
#>  class         7.3-17  2020-04-26 [4] CRAN (R 4.0.0)
#>  classInt      0.4-3   2020-04-07 [1] CRAN (R 4.0.2)
#>  cli           2.2.0   2020-11-20 [1] CRAN (R 4.0.3)
#>  colorspace    2.0-0   2020-11-11 [1] CRAN (R 4.0.3)
#>  crayon        1.3.4   2017-09-16 [1] CRAN (R 4.0.2)
#>  curl          4.3     2019-12-02 [1] CRAN (R 4.0.2)
#>  DBI           1.1.0   2019-12-15 [1] CRAN (R 4.0.2)
#>  desc          1.2.0   2018-05-01 [1] CRAN (R 4.0.2)
#>  devtools      2.3.2   2020-09-18 [1] CRAN (R 4.0.2)
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.0.3)
#>  dplyr       * 1.0.2   2020-08-18 [1] CRAN (R 4.0.2)
#>  e1071         1.7-4   2020-10-14 [1] CRAN (R 4.0.3)
#>  ellipsis      0.3.1   2020-05-15 [1] CRAN (R 4.0.2)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.0.2)
#>  fansi         0.4.1   2020-01-08 [1] CRAN (R 4.0.2)
#>  farver        2.0.3   2020-01-16 [1] CRAN (R 4.0.2)
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.0.2)
#>  generics      0.1.0   2020-10-31 [1] CRAN (R 4.0.3)
#>  ggplot2     * 3.3.3   2020-12-30 [1] CRAN (R 4.0.3)
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.0.2)
#>  gtable        0.3.0   2019-03-25 [1] CRAN (R 4.0.2)
#>  highr         0.8     2019-03-20 [1] CRAN (R 4.0.2)
#>  htmltools     0.5.0   2020-06-16 [1] CRAN (R 4.0.2)
#>  httr        * 1.4.2   2020-07-20 [1] CRAN (R 4.0.2)
#>  KernSmooth    2.23-18 2020-10-29 [4] CRAN (R 4.0.3)
#>  knitr         1.30    2020-09-22 [1] CRAN (R 4.0.2)
#>  lifecycle     0.2.0   2020-03-06 [1] CRAN (R 4.0.2)
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.0.3)
#>  memoise       1.1.0   2017-04-21 [1] CRAN (R 4.0.2)
#>  mime          0.9     2020-02-04 [1] CRAN (R 4.0.2)
#>  munsell       0.5.0   2018-06-12 [1] CRAN (R 4.0.2)
#>  pillar        1.4.7   2020-11-20 [1] CRAN (R 4.0.3)
#>  pkgbuild      1.2.0   2020-12-15 [1] CRAN (R 4.0.3)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.0.2)
#>  pkgload       1.1.0   2020-05-29 [1] CRAN (R 4.0.2)
#>  prettyunits   1.1.1   2020-01-24 [1] CRAN (R 4.0.2)
#>  processx      3.4.5   2020-11-30 [1] CRAN (R 4.0.3)
#>  ps            1.5.0   2020-12-05 [1] CRAN (R 4.0.3)
#>  purrr       * 0.3.4   2020-04-17 [1] CRAN (R 4.0.2)
#>  R6            2.5.0   2020-10-28 [1] CRAN (R 4.0.3)
#>  Rcpp          1.0.5   2020-07-06 [1] CRAN (R 4.0.2)
#>  remotes       2.2.0   2020-07-21 [1] CRAN (R 4.0.2)
#>  rlang         0.4.10  2020-12-30 [1] CRAN (R 4.0.3)
#>  rmarkdown     2.6     2020-12-14 [1] CRAN (R 4.0.3)
#>  rprojroot     2.0.2   2020-11-15 [1] CRAN (R 4.0.3)
#>  scales        1.1.1   2020-05-11 [1] CRAN (R 4.0.2)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.0.2)
#>  sf          * 0.9-7   2021-01-06 [1] CRAN (R 4.0.3)
#>  stringi       1.5.3   2020-09-09 [1] CRAN (R 4.0.2)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.0.2)
#>  testthat      3.0.1   2020-12-17 [1] CRAN (R 4.0.3)
#>  tibble        3.0.4   2020-10-12 [1] CRAN (R 4.0.3)
#>  tidyselect    1.1.0   2020-05-11 [1] CRAN (R 4.0.2)
#>  units         0.6-7   2020-06-13 [1] CRAN (R 4.0.2)
#>  usethis       2.0.0   2020-12-10 [1] CRAN (R 4.0.3)
#>  utf8          1.1.4   2018-05-24 [1] CRAN (R 4.0.2)
#>  vctrs         0.3.6   2020-12-17 [1] CRAN (R 4.0.3)
#>  withr         2.3.0   2020-09-22 [1] CRAN (R 4.0.2)
#>  xfun          0.20    2021-01-06 [1] CRAN (R 4.0.3)
#>  xml2          1.3.2   2020-04-23 [1] CRAN (R 4.0.2)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.0.2)
#> 
#> [1] /home/floris/lib/R/library
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

</details>

**Comment by hansvancalster** [comment]
> Any further ideas for the remaining cases? I'm prepared to post an issue at `sf` for this: extending `st_cast()`, perhaps also requesting out-of-the-box support for extra feature types, as has been done in [r-spatial/sf@fe70fea](https://github.com/r-spatial/sf/commit/fe70feab8375147bfa4c8299496785051c7f08e7) for `MULTICURVE`, IIUC.

I agree that posting an issue in the `sf` package is the best way to proceed for these exotic geometry types. Support in `st_cast()` should be sufficient.

@florisvdh do you have remaining comments? If not, can you approve the PR for now?

**Review by florisvdh — APPROVED** [review]
Thanks for the updates @hansvancalster .

Please re-knit after running locally `remotes::install_github("eblondel/ows4R")` :wink: . That will remove the warning from the md file:

```
## Warning in CPL_crs_from_input(x): GDAL Message 1: +init=epsg:XXXX syntax is
## deprecated. It might return a CRS with a non-EPSG compliant axis order.
```

When looking at the rendered result (nice buffer picture!), the example4 link did not work:

afbeelding

I propose updates in two places for that link, and that result will work. (after re-knitting)

Finally:

> Some of the material presented in this tutorial benefitted from

Here you could perhaps add the Dutch wordpress blogpost https://friesewoudloper.wordpress.com/2015/05/17/het-bevragen-van-een-wfs-in-de-browser-voorbeelden-aan-de-hand-van-de-bag-service/

**Comment by florisvdh** [comment]
> I agree that posting an issue in the sf package is the best way to proceed for these exotic geometry types. Support in st_cast() should be sufficient.

Continued in r-spatial/sf#1573

File: content/tutorials/spatial_wfs_services/index.Rmd [hunk 2]
```diff
@@ -300,20 +311,41 @@ wfs_vrbg %>%
...
+  st_buffer(dist = 10000) # works
+
+sf_prov %>% 
+  ggplot() + 
+  geom_sf(data = sf_prov_buffer) +
+  geom_sf()
 ```
 
 
+Also check-out [example 4](example4) for a more advanced use of the CQL filter.
```
[the remark this answers is not in the document]
Comment (florisvdh) [reply content/tutorials/spatial_wfs_services/index.Rmd:348]: ignore this one; it is included (and then moved) in PR #220

**Comment by florisvdh** [comment]
r-spatial/sf#1573 is been solved in the GitHub version of **sf** !! :tada: 

@hansvancalster some final suggestions are in PR #220 (which points to this branch), after further looking into those exotic geometry types. Still to be knitted!

**Comment by hansvancalster** [comment]
> Please re-knit after running locally `remotes::install_github("eblondel/ows4R")`  . That will remove the warning from the md file:

Strange, the warning should have been gone. I had already installed that version of the package.

**Comment by florisvdh** [comment]
> Strange, the warning should have been gone. I had already installed that version of the package.

That may be in another place, I saw it here:

afbeelding

**Comment by hansvancalster** [comment]
> That may be in another place, I saw it here (confirmed, it's still there currently):

Indeed :thinking: When I knit it, it is there (but in another place as before). When I run the chunks, the warning isn't there.

**Pull request merged**